completions/xrandr

```diff
@@ -1,62 +1,165 @@
 # bash completion for xrandr                               -*- shell-script -*-
 
+_xrandr_outputs()
+{
+    "$1" 2>/dev/null | awk '/connected/ {print $1}'
+}
+
+_xrandr_monitors()
+{
+    "$1" --listmonitors 2>/dev/null | sed -ne 's/.* [0-9]*: [+\*]*\([^ ]*\).*/\1/p'
+}
+
+_xrandr_providers()
+{
+    "$1" --listproviders 2>/dev/null | sed -ne "s/.*cap:.*$2.* name:\([^ ]*\).*/\1/p"
+}
+
+_xrandr_modes()
+{
+    "$1" -q 2>/dev/null | sed -e "1,/^$2 / d" \
+                              -e '/connected/,$ d' \
+                              -e '/^[[:space:]]*h: / d' \
+                              -e '/^[[:space:]]*v: / d' \
+                              -e 's/\([^[:space:]]\)[[:space:]].*/\1/'
+}
+
+_xrandr_help()
+{
+    "$1" --help 2>/dev/null
+}
+
 _xrandr()
 {
     local cur prev words cword
     _init_completion || return
 
+    local i output
+    for ((i = cword - 1; i > 0; i--)); do
+        if [[ ${words[i]} == --output ]]; then
+            output=${words[i + 1]}
+            break
+        fi
+    done
+
     case "$prev" in
-        -display | -d | -help | -s | --size | -r | --rate | --refresh | --screen | --fb | --fbmm | \
-            --dpi | --pos | --set | --scale | --transform | --crtc | --panning | --gamma | \
-            --newmode | --rmmode | --addmode | --delmode)
+        -display | -d | -s | --size | -r | --rate | --refresh | --screen | --fb | --fbmm | \
+            --pos | --set | --scale | --transform | --crtc | --panning | --gamma | \
+            --newmode | --rmmode | --setmonitor)
+            return
+            ;;
+        --output | --addmode | --delmode)
+            local outputs=$(_xrandr_outputs "$1")
+            COMPREPLY=($(compgen -W "$outputs" -- "$cur"))
             return
             ;;
-        --output | --left-of | --right-of | --above | --below | --same-as)
-            local outputs=$("$1" | awk '/connected/ {print $1}')
+        --dpi)
+            local outputs=$(_xrandr_outputs "$1")
             COMPREPLY=($(compgen -W "$outputs" -- "$cur"))
```

Comment:
Hmm, this doesn't look quite right. Completing `--dpi` with outputs?

Reply:
Yes, this is strange, but both the man page, the help text, and the [source code](https://gitlab.freedesktop.org/xorg/app/xrandr/-/blob/master/xrandr.c#L3125) of `xrandr` indicate that the name of an output is an acceptable argument.

Reply:
Ok :P In that case, I suggest combining it with the identical `--output | --addmode | --delmode` switch case or adding a comment why it's a separate one.

Reply:
Done.